resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -332,16 +335,116 @@ private InetAddress resolveHostsFileEntry(String hostname) {
+     * @param additionals additional records ({@code OPT})
+     * @param promise the {@link Promise} which will be fulfilled when the name resolution is finished
+     *
+     * @return the address as the result of the resolution
+     */
+    public final Future<InetAddress> resolve(String inetHost, Iterable<DnsRecord> additionals,
+                                             Promise<InetAddress> promise) {
+        checkNotNull(inetHost, "inetHost");
+        checkNotNull(additionals, "additionals");
+        checkNotNull(promise, "promise");
+
+        try {
+            doResolve(inetHost, validateAdditionals(additionals), promise, resolveCache);
+            return promise;
+        } catch (Exception e) {
+            return promise.setFailure(e);
+        }
+    }
+
+    /**
+     * Resolves the specified host name and port into a list of address.
+     *
+     * @param inetHost the name to resolve
+     * @param additionals additional records ({@code OPT})
+     *
+     * @return the list of the address as the result of the resolution
+     */
+    public final Future<List<InetAddress>> resolveAll(String inetHost, Iterable<DnsRecord> additionals) {
+        return resolveAll(inetHost, additionals, executor().<List<InetAddress>>newPromise());
+    }
+
+    /**
+     * Resolves the specified host name and port into a list of address.
+     *
+     * @param inetHost the name to resolve
+     * @param additionals additional records ({@code OPT})
+     * @param promise the {@link Promise} which will be fulfilled when the name resolution is finished
+     *
+     * @return the list of the address as the result of the resolution
+     */
+    public final Future<List<InetAddress>> resolveAll(String inetHost, Iterable<DnsRecord> additionals,
+                                                Promise<List<InetAddress>> promise) {
+        checkNotNull(inetHost, "inetHost");
+        checkNotNull(additionals, "additionals");
+        checkNotNull(promise, "promise");
+
+        try {
+            doResolveAll(inetHost, additionals, promise, resolveCache);
+            return promise;
+        } catch (Exception e) {
+            return promise.setFailure(e);
+        }
+    }
+
     @Override
     protected void doResolve(String inetHost, Promise<InetAddress> promise) throws Exception {
-        doResolve(inetHost, promise, resolveCache);
+        doResolve(inetHost, null, promise, resolveCache);
+    }
+
+    private static DnsRecord[] validateAdditionals(Iterable<DnsRecord> additionals) {
+        if (additionals instanceof Collection) {
+            Collection<DnsRecord> records = (Collection<DnsRecord>) additionals;
+            for (DnsRecord r: additionals) {
+                validateAdditional(r);
+            }
+            return records.toArray(new DnsRecord[records.size()]);
+        }
+
+        Iterator<DnsRecord> additionalsIt = additionals.iterator();
+        if (!additionalsIt.hasNext()) {
+            return null;
```

Comment:
Could we return an empty singleton array and remove the null check in `emptyAdditionals()`?